### ISSUE TITLE
[Lite] Assertion failure if shape of dynamic output tensor changes between invoke()s

### ISSUE THREAD

**Issue by scottcjt** [tensorflow/tensorflow#26549]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 1.13.1 (cpu)
- Python version: 3.6.7
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

**Moved from #26248.**

If an op:

* depends on a (1) dynamic tensor, and (2) a normal intermediate tensor
* the computation of (2) is performed before (1)
* dims of (1) change between `Interpreter::invoke()`s

Then assertion failure `ERROR: tensorflow/lite/simple_memory_arena.cc:100 erased_allocs_count != 1 (0 != 1)` is triggered.

An unit test is attached in #26248. On my box it is also reproducible by the Python snippet below, which generate a model dynamically. However, my reviewer can only reproduce the issue by unit test, but not by script, so the script may miss something. To help debugging, a TFLite model file which I can reproduce the issue with is uploaded: [failed.tar.gz](https://github.com/tensorflow/tensorflow/files/2949840/failed.tar.gz)

```python
import numpy as np
import tensorflow as tf


i = tf.placeholder(name='i', dtype=tf.float32, shape=[1])
p = tf.placeholder(name='p', dtype=tf.int32, shape=[1, 2])
a = tf.placeholder(name='a', dtype=tf.float32, shape=[1])

n = tf.negative(a)
o = tf.add(tf.pad(i, p), n)

test_input = np.array([0], dtype=np.float32)
test_pad = np.array([[2,2]], dtype=np.int32)
test_pad_2 = np.array([[4,4]], dtype=np.int32)
test_a = np.array([1], dtype=np.float32)


def test_tf():
    with tf.Session() as sess:
        out = sess.run(o, {i: test_input, p: test_pad, a: test_a})
        assert out.shape == (5,)

        out = sess.run(o, {i: test_input, p: test_pad_2, a: test_a})
        assert out.shape == (9,)

test_tf()


# convert
with tf.Session() as sess:
    conv = tf.lite.TFLiteConverter.from_session(sess, [i, p, a], [o])
    lite_model_bytes = conv.convert()

    # with open('failed.tflite', mode='wb') as fp:
    #     fp.write(lite_model_bytes)


def test_tflite():
    interp = tf.lite.Interpreter(model_content=lite_model_bytes)

    #### I just hard-coded them:
    # print(interp.get_input_details())
    # print(interp.get_output_details())

    interp.allocate_tensors()
    interp.set_tensor(4, test_input)
    interp.set_tensor(5, test_pad)
    interp.set_tensor(3, test_a)

    interp.invoke()
    assert interp.get_tensor(0).shape == (5,)

    interp.set_tensor(5, test_pad_2)

    interp.invoke()
    assert interp.get_tensor(0).shape == (9,)

test_tflite()

```

**Comment by haozha111** [comment]
Hi Scott,

Sorry for late reply! Can you still reproduce this issue? I tested it on tf-nightly but no assertion errors show up again.

**Comment by scottcjt** [comment]
The unit test added in #26248 still failed with latest commit on master (b0432d52). As for the python script, it is not always reproducible.

**Comment by ANSHUMAN87** [comment]
@scottcjt : The issue is resolved in Tensorflow 1.14. Would you please check and confirm.
If still you are facing issue, please let me know, Thanks!

**Comment by umangmehta12** [comment]
@ANSHUMAN87 can you point me to the specific location where this is fixed in 1.14? I am facing similar issue with 1.14

**Comment by ANSHUMAN87** [comment]
@umangmehta12 : Are you able to reproduce the issue in 1.14, using the scripts provided by the author of this issue? There is big difference most of the time between similar and same issue. If not able to reproduce the issue with the script, then try to raise a new issue with your reproducible script and issue details. I can help you. You can TAG me if you are raising a new one.
